### PR TITLE
Update javascript to typescript

### PR DESCRIPTION
**Pull request by ClemClark258**
Updated Javascript files for SupplierManagement and CatalogueSolutionsFilter to conform with Typescript standards

### CODE REVIEW

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/js/CatalogueSolutionsFilter.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
+const filterUrl = "/filter";
```
Comment (bjss-colin) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/js/CatalogueSolutionsFilter.js:1]: It might be worth adding a wildcard to the gitignore to ignore these files as they're generated at build time.

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    "configurations": [
+        {
+            "type": "pwa-chrome",
+            "request": "launch",
+            "name": "Launch Chrome against localhost",
+            "url": "http://localhost:5001",
+            "webRoot": "${workspaceFolder}"
+        }
+    ]
+}
```
Comment (bjss-chris) [inline .vscode/launch.json:15]: We shouldn't be committing files in the .vscode directory.

We probably ought to add `.vscode/*` to `gitignore` as VS Code is generally nicer to work with for JS/TS.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "gulp": "^4.0.2"
+  }
+}
```
Comment (bjss-chris) [inline package.json:5]: As we already have `package.json` and `package.json` in the `WebApp` directory we shouldn't need them here, so recommend deleting both files from this directory.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Views/Solutions/Index.cshtml [hunk 4]
```diff
@@ -35,7 +35,7 @@
 }
 
 @section Scripts{ 
-    <script type="text/javascript" src="@Url.Content("~/js/CatalogueSolutionsFilter.min.js")" asp-append-version="true"></script>
+    <script type="text/javascript" src="@Url.Content("~/js/CatalogueSolutionsFilter.js")" asp-append-version="true"></script>
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Views/Solutions/Index.cshtml:38]: We should still reference the minified version here.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/gulpfile.js [hunk 5]
```diff
@@ -6,7 +6,7 @@ const gulp = require("gulp"),
     terser = require('gulp-terser'),
     merge = require("merge-stream"),
     del = require("del"),
-    sass = require("gulp-sass")(require('sass')),
+    sass = require("gulp-sass")(require("sass")),
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/gulpfile.js:9]: This change shouldn't be required technically but guess it makes sense to be consistent 👍

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/package.json [hunk 6]
```diff
@@ -8,21 +8,29 @@
   },
   "keywords": [],
   "author": "",
-  "license": "ISC",
   "dependencies": {
     "govuk-frontend": "^3.13.1",
+    "gulp-terser": "^2.1.0",
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/package.json:13]: We shouldn't need `gulp-terser` as a production dependency as the only time we run `gulp` is at build time.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/package.json [hunk 7]
```diff
@@ -8,21 +8,29 @@
...
-  }
+  },
+  "keywords": [],
+  "license": "ISC",
+  "main": "index.js",
+  "name": "NHSD.GPIT.BuyingCatalogue.WebApp",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "version": "1.0.0"
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/package.json:35]: Isn't this info already present (at the start of the file)?

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/tsconfig.json [hunk 8]
```diff
@@ -1,19 +1,20 @@
...
+    "noImplicitAny": true,
     "noEmitOnError": true,
-    "removeComments": false,
+    "removeComments": true,
     "sourceMap": true,
-    "target": "es6",
+    "target": "es2019",
     "outDir": "wwwroot/js",
-    "rootDir": "./Scripts",
+    "rootDir": "wwwroot/ts",
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/tsconfig.json:11]: I think it makes sense for the `.ts` scripts to stay where they were in the `Scripts` directory.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/tsconfig.json [hunk 9]
```diff
@@ -1,19 +1,20 @@
...
     "lib": [
       "es2019",
       "dom"
     ]
   },
   "include": [
-    "Scripts/**/*"
+    "wwwroot/ts/*"
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/tsconfig.json:18]: As above.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+        this.RefireDomContentLoadedEvent();
+
+        this.reselectCapabilityAndEpicsFiltersAndFrameworkFilter();
+    }
+
+    generateQueryParam=() => {
+        const queryPage: string = "page";
+        let output: string = "";
+        let query: string = "#" + filterContainer + " " + NhsukCheckboxesInput;
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:51]: Can we make this `const`?

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+        document.querySelectorAll(query).forEach(checkbox => {
+            const input: HTMLInputElement = checkbox as HTMLInputElement;
+            if (input.checked) {
+
+                let value: string = null;
+                value = checkbox.parentNode!.querySelector(InputTypeHidden)!.getAttribute("value");
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:58]: `const`?

```ts
const value: string = checkbox.parentNode!.querySelector(InputTypeHidden)!.getAttribute("value");
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+
+                if (value.includes(SupplierSplitCharacter)) {
+                    output += this.EncodeSupplierDefinedEpic(value);
+                } else if (value.includes(EpicSplitCharacter) && !value.includes(CapabilityMarker)) {
+                    output += this.EncodeDFOCVCEpic(value);
+                } else if (value.includes(EpicSplitCharacter) && value.includes(CapabilityMarker)) {
+                    output += this.EncodeNormalEpic(value);
+                } else if (output.length > 1) {
+                        output += CapabilitiesDelimiter + value;
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:67]: Looks like this is tabbed in one indent too far.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+                } else {
+                    output += value;
+                }
+            }
+        });
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:72]: This section here looks like it could probably be extracted into a separate function. However, that might be worth leaving for a separate PR.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+            const input: HTMLInputElement = radioInputs[i] as HTMLInputElement;
+
+            if (input.checked) {
+                selectedFramework = input.value;
+                break;
+            }
+        }
+
+        return selectedFramework;
+    }
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:182]: Rather than having the mutable variable, `selectedFramework`, can we do something like this here?

```ts
GetSelectedFramework(): string {
    const radioInputs: NodeListOf<Element> = document.querySelectorAll(NhsRadiosInput);
    
    for (let i : number = 0; i < radioInputs.length; i++) {
        const input: HTMLInputElement = radioInputs[i] as HTMLInputElement;

        if (input.checked) {
            return input.value;
        }
    }

    return 'All';
}
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+
+    RefireDomContentLoadedEvent(): void {
+        window.document.dispatchEvent(new Event("DOMContentLoaded", {
+            bubbles: true,
+            cancelable: true,
+        }));
+    }
+
+
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:191]: Surplus white space here – one line should do the job 😃

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+};
+
+function refreshCapabilitiesAndEpics(html: string): void {
+    const filterHtml: Document = new DOMParser().parseFromString(html, "text/html");
+
+    let newCapabilities: HTMLElement = null;
+    newCapabilities = filterHtml.getElementById(filterCapabilitiesDetails)!.querySelector(filterDetailsText);
+
+    let currentCapabilities: HTMLElement = null;
+    currentCapabilities = document.getElementById(filterCapabilitiesDetails)!.querySelector(filterDetailsText);
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:273]: Can we make these two vars `const`? E.g.

```ts
const newCapabilities: HTMLElement = filterHtml.getElementById(filterCapabilitiesDetails)!.querySelector(filterDetailsText);
const currentCapabilities: HTMLElement = document.getElementById(filterCapabilitiesDetails)!.querySelector(filterDetailsText);
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/SupplierManagement.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
+﻿
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/SupplierManagement.ts:1]: Surplus white space.

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts [hunk 18]
```diff
@@ -0,0 +1,280 @@
+﻿const filterUrl: string = "/filter";
```
[the remark this answers is not in the document]
Comment (bjss-colin) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/CatalogueSolutionsFilter.ts:1]: Marking as resolved on that basis

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/SupplierManagement.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    ChangeSupplierDisplay(): void {
+
+        this.checked ? ChangeSupplierDisplayStyle("table-row") : ChangeSupplierDisplayStyle("none");
+    }
+}
+
+window.onload = function (): void {
+    var supplierManagement: SupplierManagement = new SupplierManagement();
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/SupplierManagement.ts:27]: Can be `const`. (Generally best to avoid `var` and use `let` if we need a mutable var).

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/SupplierManagement.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+function ChangeSupplierDisplayStyle(style : string): void {
+    const inactiveSuppliers : HTMLCollectionOf<Element> = document.getElementsByClassName("inactive");
+    for (let i : number = 0; i < inactiveSuppliers.length; i++) {
+        const input : HTMLInputElement = inactiveSuppliers.item(i) as HTMLInputElement;
+        input.style.display = style;
+    }
+}
+
+
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/wwwroot/ts/SupplierManagement.ts:39]: Surplus whitespace (just need the one trailing new line).

**Review by bjss-chris — CHANGES_REQUESTED** [review]
There are a lot of files included in the commit that don't need to be, particular from the nhsuk-frontend library (e.g. all the `.png` files), minified JS files and `package files` that should be removed.

Like @bjss-colin, I think the `.ts` files would be better remaining in the `Scripts` directory.

I've left a few comments on the TS (mostly around making use of `const`) but the actual TS scripts look alright on the whole though! 👍
